### DISCUSSION TITLE
Apply `bound_emission` to all nodes under a parent node?

### DISCUSSION THREAD

**Discussion by SongminYu** [iiasa/message_ix#845] [Q&A] [answered]
Hi MESSAGE team and everyone, hope my message finds you all well!

I am trying to follow this [emission bound](https://github.com/iiasa/message_ix/blob/main/tutorial/westeros/westeros_emissions_bounds.ipynb) example in a bit different setup where region hierarchy exists. Taking the EU and members states as example, due to data limitation,

- I have the EU as a node holding all demand-side sectors and their technologies
- I also have other member states (e.g., Austria, Germany, Finland, etc.) as nodes in the model, but only with power plants. All these power plants have their 
  - input of coal, gas, etc. from the node EU 
  - input of solar and wind from their own regions
  - output of electricity goes to the EU again
- In the EU node, I have for example a cement sector which has CO2 emission. 

Now, I want to apply a CO2 emission bound to the whole EU. From the following code,

```python
scen.add_par(
    "bound_emission",
    make_df(
        "bound_emission",
        node=country,
        type_emission="GHG",
        type_tec="all",
        type_year="cumulative",
        value=500.0,
        unit="MtCO2",
    ),
)
```

It seems that, since the index is `node` not `type_node`, the CO2 emission bound is only applied to the cement sector in the EU (node) not the power plants in the member states? If so, my question is, is it possible to edit the region hierarchy so that the model knows all the CO2 in the model should be constrained? Or, is there another way around?

Thanks!

**Comment by khaeru** [comment, marked as answer]
@SongminYu this is a good question. Not a complete answer, but three points of information.

1. There is a distinction between:

   - (a) The "node" concept as used in message_ix as a set and dimension on several parameters,
   - (b) The "region" concept used by the underlying ixmp storage layer for IAMC-structured (“time-series”) data.

   In particular:

   - (a) are specific to a single Scenario, and may differ entirely across multiple Scenarios on the same Platform. Hierarchy information can *not* be represented.
   - (b) are common to all time-series/IAMC-structured data on the same Platform. These *can* be used to store and retrieve hierarchy information.

   To be clear, adding codes/elements to the "node" set in one Scenario *does not* affect the "regions" on the Platform, and vice versa. They are not necessarily connected in any way, except that they represent something that is conceptually similar.

   This has been baked-in to the underlying data model of ixmp for years, and we recognize that it often causes confusion.

2. Your objective may be similar to #514. Note that:
   - That PR is in "draft" state and is about 3 years old. Its code may not be compatible with current versions of ixmp and message_ix.
   - Nevertheless, it contains changes to the GAMS code that you *could* attempt to use on your system. I would recommend trying this by applying the diff/patch to the latest GAMS files.
   - It may not work, but if you try this, please do correspond with the PR authors by leaving a comment to report your experience.

3. Another strategy may be to use the ["generic relations" parameters and equations](https://docs.messageix.org/en/latest/model/MESSAGE/parameter_def.html#parameters-of-the-relations-section) of `message_ix`. The approach would be roughly as follows:

   - Add "pseudo-technologies" in your `node="EU"` using codes like, for instance, `technology="coal_ppl_emi_AT"`. Do not add any techno-economic parameter data (`*_cost`, `input`, or `output`, etc.) for these technologies.
   - Use parameters including `relation_activity` and related sets to establish "relation" constraints such that the activity (`ACT`) of (t=coal_ppl_emi_AT, n=EU) is equal to the `ACT` of (t=coal_ppl, n=AT). For instance, you could give one a coefficient (`relation_activity` parameter value) of 1.0, the other a coefficient of -1.0, and then set `relation_lower = relation_upper = 0.0`.
   - Add `emission_factor` data for (t=coal_ppl_emi_AT, n=EU). *Remove* (or do not add) `emission_factor` data for (t=coal_ppl, n=AT).

   With this approach:

   1. (t=coal_ppl, n=AT) will be active to supply the `demand` you have parametrized. This activity incurs costs, consumes inputs, produces outputs, affects the objective function value (`OBJ`), etc.
   2. The `relation_activity` values force an equal activity for  (t=coal_ppl_emi_AT, n=EU). This is equivalent to $$0.0 \leq (1.0 \times ACT_{t=coal.ppl.emi.AT, n=EU} + (-1.0) \times ACT_{t=coal.ppl, n=AT}) \leq 0.0$$.
   3. (t=coal_ppl_emi_AT, n=EU), via its `emission_factor`, contributes to `EMISS` that must be, in aggregate, below the bound you set for n=EU. This also affects the `OBJ`. But this "pseudo-technology" has no costs *per se*, does not consume or produce anything, etc. It exists only to represent emissions occurring in n=EU.
   4. The balance of marginal contributions to `OBJ` of (1) and (3) is functionally equivalent to having the techno-economic parameters *and* `emission_factor` applying to the same (n, t).

   Note that if `emission_factor` will be have the same values for t=coal_ppl in all your nodes, you could simplify and use a single pseudo-technology like t=coal_ppl_emi in n=EU.

**Reply by SongminYu** [reply]
Hi @khaeru, thanks a lot for your swift and structured answer! Super helpful! I somehow had the feeling that "generic relations" could be useful in this situation. What you explained gives me a much clearer idea of it. 

In the end, I realized that I could do something even easier: I assigned the `emission_factor` to the `coal_import` and `gas_import` technologies which are located at node EU. Because I only have the power sector modeled at the "member states" level, to capture the heterogenous renewable potential. All the coal or gas they use comes from the node EU and they don't have "process emission" like cement production. Luckily, in my current setup, all the complicated demand-side sectors (e.g., cement with process emission) are already located at the node EU. So, I can limit the EU emission only now. 

I know it may not be accurate enough and conceptually it is not clean, but I decided to go with this at the moment. In the next stage, if necessary, I will consider using the generic relations to refactor it. If, by that time, the "emission pool" is solved, that would be even better :)

Thanks again for your answer!

**Reply by behnam-zakeri** [reply]
@SongminYu and @khaeru thanks for the discussion. If you would like to calculate the emissions of a node, in this case EU, including the nodes under that node, in this case EU member states, you can do that more easily by using spatial hierarchy in `message_ix`. This will allow you to use `bound_emission` straightforward. @SongminYu did you tell the model that your EU node is the "parent" node of the country node? You should do that using MESSAGEix set `map_spatial_hierarchy`. Then, the model automatically calculates emissions for EU, accounting for all the emissions from the "child" nodes.
The coding is relatively straightforward:
1. Introduce different spatial levels for EU and member states. For example, "country" for member states and "region" for EU:
`scen.add_set("lvl_spatial", ["country", "region"])`
2. Then, tell the model which node is at which level, and what the parent node is for that node, e.g.:
```
scen.add_set("map_spatial_hierarchy", ["region", "EU", "World"])
for node in member_states:
    scen.add_set("map_spatial_hierarchy", ["country", node, "EU"])
```
That's it. Here `member_state` is the list of your country nodes to be added under "EU". Then you can add `bound_emission` only for "EU" as you showed in your question.

**Reply by khaeru** [reply]
@behnam-zakeri thanks for filling in on this model feature that I wasn't really aware of!

As an aside, I think this points to the need for a 'space' docs page parallel to https://docs.messageix.org/en/latest/time.html, to explain such topics.

**Reply by behnam-zakeri** [reply]
I agree @khaeru. This should be more prominent in the documentation.

**Reply by SongminYu** [reply]
Hi @behnam-zakeri, thanks for the reply! I just tried implementing the `lvl_spatial` logic. It works perfectly! I see the emissions aggregated at the top node in `EMISS` output.
